### PR TITLE
Estimate BETWEEN when the value is an expression

### PR DESCRIPTION
**Pull request by raunaqmorarka**
These changes will result in symbol estimates only in CAST(symbol) case
and row count estimate in more general cases.

### CODE REVIEW

File: core/trino-main/src/main/java/io/trino/cost/FilterStatsCalculator.java [hunk 1]
```diff
@@ -396,6 +397,17 @@ protected PlanNodeStatsEstimate visitFunctionCall(FunctionCall node, Void contex
...
 
+        private Optional<Symbol> getSymbolReference(Expression expression)
+        {
+            if (expression instanceof SymbolReference) {
+                return Optional.of(Symbol.from(expression));
+            }
+            if (!(expression instanceof Cast) || !(((Cast) expression).getExpression() instanceof SymbolReference)) {
+                return Optional.empty();
+            }
+            return Optional.of(Symbol.from(((Cast) expression).getExpression()));
```
Comment (findepi) [inline core/trino-main/src/main/java/io/trino/cost/FilterStatsCalculator.java:408]: unwrapping Cast here looks like an assumption that "stats representation" (eg ranges) before and after cast are comparable. I.e. that `io.trino.spi.statistics.StatsUtil#toStatsRepresentation` is the same (or almost the same) before and after the cast. 

This isn't true, however, is it?

File: core/trino-main/src/main/java/io/trino/cost/FilterStatsCalculator.java [hunk 2]
```diff
@@ -371,7 +372,7 @@ protected PlanNodeStatsEstimate visitComparisonExpression(ComparisonExpression n
             }
 
             SymbolStatsEstimate leftStats = getExpressionStats(left);
-            Optional<Symbol> leftSymbol = left instanceof SymbolReference ? Optional.of(Symbol.from(left)) : Optional.empty();
+            Optional<Symbol> leftSymbol = getSymbolReference(left);
```
Comment (sopel39) [inline core/trino-main/src/main/java/io/trino/cost/FilterStatsCalculator.java:375]: I'm not sure it's right thing to do. For example `io.trino.cost.ComparisonStatsCalculator#estimateFilterRange` will set column statistics for that symbol based on cast `leftStats` (please correct me if I'm wrong). It could be that `CAST` is narrowing down, so that target domain stats don't make sense for original symbol.

If this code brings benefits, then I would do either:
* add TODO to `getSymbolReference` to somehow limit unwrapping of `CAST` to cases where target domains is wider or equal
* make `getSymbolReference` unwrap `CAST` to to cases where target domains is wider or equal

**Comment by sopel39** [comment]
make sure this improves queries before it gets merged.

**Comment by raunaqmorarka** [comment]
> make sure this improves queries before it gets merged.

I can benchmarks for the small subset of queries where this scenario actually arises. It seems there are small gains
Unpartitioned ORC
between unpartitioned

Partitioned ORC
between partittioned

File: core/trino-main/src/main/java/io/trino/cost/FilterStatsCalculator.java (hunk 1 above)
Comment (raunaqmorarka) [reply core/trino-main/src/main/java/io/trino/cost/FilterStatsCalculator.java]: Yes, I've now added logic to do this unwrapping only if the cast is `isTypeOnlyCoercion`
PTAL and let me know if this is narrow enough to ensure correctness

File: core/trino-main/src/main/java/io/trino/cost/FilterStatsCalculator.java (hunk 2 above)
Comment (raunaqmorarka) [reply core/trino-main/src/main/java/io/trino/cost/FilterStatsCalculator.java]: I've now added logic to do this unwrapping only if the cast is `isTypeOnlyCoercion`.
I'm assuming this will ensure that the CAST is to a wider domain

Comment (findepi) [reply core/trino-main/src/main/java/io/trino/cost/FilterStatsCalculator.java]: > I've now added logic to do this unwrapping only if the cast is isTypeOnlyCoercion.

`isTypeOnlyCoercion`  is about Trino stack representation.
This is not the same as "stats representation", the two meanings should not be conflated.
At the very least you should have a 
- a method in `StatsUtil` that tells which type pairs (which casts)  it's safe to unwrap
- have a test that verifies `StatsUtil.toStatsRepresentation` is the same before and after the cast, for the types allowed for unwrapping

Comment (raunaqmorarka) [reply core/trino-main/src/main/java/io/trino/cost/FilterStatsCalculator.java]: I've added `StatsUtil#areStatsRepresentationsCompatible` now and some testing around it in `FilterStatsCalculator#testUnwrapCastSymbolReference`

File: core/trino-spi/src/main/java/io/trino/spi/statistics/StatsUtil.java [hunk 3]
```diff
@@ -85,4 +85,44 @@ public static OptionalDouble toStatsRepresentation(Type type, Object value)
 
         return OptionalDouble.empty();
     }
+
+    public static boolean areStatsRepresentationsCompatible(Type type, Type otherType)
```
Comment (findepi) [inline core/trino-spi/src/main/java/io/trino/spi/statistics/StatsUtil.java:89]: Document

File: core/trino-spi/src/main/java/io/trino/spi/statistics/StatsUtil.java [hunk 4]
```diff
@@ -85,4 +85,44 @@ public static OptionalDouble toStatsRepresentation(Type type, Object value)
...
+        requireNonNull(type, "type is null");
+        requireNonNull(otherType, "otherType is null");
+
+        if (type.equals(otherType)) {
+            return true;
+        }
+
+        if (type == BOOLEAN || type == REAL || type == DOUBLE || type == DATE) {
+            return false;
+        }
```
Comment (findepi) [inline core/trino-spi/src/main/java/io/trino/spi/statistics/StatsUtil.java:100]: Redundant.

File: core/trino-spi/src/main/java/io/trino/spi/statistics/StatsUtil.java [hunk 5]
```diff
@@ -85,4 +85,44 @@ public static OptionalDouble toStatsRepresentation(Type type, Object value)
...
+        if (type == TINYINT || type == SMALLINT || type == INTEGER || type == BIGINT) {
+            return (otherType == TINYINT || otherType == SMALLINT || otherType == INTEGER || otherType == BIGINT);
```
Comment (findepi) [inline core/trino-spi/src/main/java/io/trino/spi/statistics/StatsUtil.java:102]: braces are redundant.

Comment (findepi) [reply core/trino-spi/src/main/java/io/trino/spi/statistics/StatsUtil.java:102]: also, `decimal(p, 0)` surely has compatible stats representation with integer types (may matter for Oracle connector, for example)

also, `decimal(p, s)` has "compatible" stats representation, depending on the definition of "compatible"

File: core/trino-spi/src/main/java/io/trino/spi/statistics/StatsUtil.java [hunk 6]
```diff
@@ -85,4 +85,44 @@ public static OptionalDouble toStatsRepresentation(Type type, Object value)
...
+        }
+        if (type instanceof DecimalType) {
+            if (!(otherType instanceof DecimalType)) {
+                return false;
+            }
+            DecimalType decimalType = (DecimalType) type;
+            DecimalType otherDecimalType = (DecimalType) otherType;
+            return (decimalType.isShort() == otherDecimalType.isShort())
```
Comment (findepi) [inline core/trino-spi/src/main/java/io/trino/spi/statistics/StatsUtil.java:110]: Why both need to be short or both long?

File: core/trino-spi/src/main/java/io/trino/spi/statistics/StatsUtil.java [hunk 7]
```diff
@@ -85,4 +85,44 @@ public static OptionalDouble toStatsRepresentation(Type type, Object value)
...
+                    && (decimalType.getScale() == otherDecimalType.getScale());
```
Comment (findepi) [inline core/trino-spi/src/main/java/io/trino/spi/statistics/StatsUtil.java:111]: For some definitions of "compatible", this `decimal(p, s)` and `decimal(p+1, s+1)` are "compatible"

File: core/trino-spi/src/main/java/io/trino/spi/statistics/StatsUtil.java [hunk 8]
```diff
@@ -85,4 +85,44 @@ public static OptionalDouble toStatsRepresentation(Type type, Object value)
...
+        }
+        if (type instanceof TimestampType) {
+            if (!(otherType instanceof TimestampType)) {
+                return false;
+            }
+            return ((TimestampType) type).isShort() == ((TimestampType) otherType).isShort();
```
Comment (findepi) [inline core/trino-spi/src/main/java/io/trino/spi/statistics/StatsUtil.java:117]: Why both need to be short or both long?

File: core/trino-spi/src/main/java/io/trino/spi/statistics/StatsUtil.java [hunk 9]
```diff
@@ -85,4 +85,44 @@ public static OptionalDouble toStatsRepresentation(Type type, Object value)
...
+        }
+        if (type instanceof TimestampWithTimeZoneType) {
+            if (!(otherType instanceof TimestampWithTimeZoneType)) {
+                return false;
+            }
+            return ((TimestampWithTimeZoneType) type).isShort() == ((TimestampWithTimeZoneType) otherType).isShort();
```
Comment (findepi) [inline core/trino-spi/src/main/java/io/trino/spi/statistics/StatsUtil.java:123]: Why both need to be short or both long?

**Review by findepi — CHANGES_REQUESTED** [review]

File: core/trino-spi/src/main/java/io/trino/spi/statistics/StatsUtil.java (hunk 5 above)
Comment (raunaqmorarka) [reply core/trino-spi/src/main/java/io/trino/spi/statistics/StatsUtil.java]: I've added compatibility for `decimal(p, 0)` with integer types and removed extra braces

File: core/trino-spi/src/main/java/io/trino/spi/statistics/StatsUtil.java (hunk 7 above)
Comment (raunaqmorarka) [reply core/trino-spi/src/main/java/io/trino/spi/statistics/StatsUtil.java:122]: I've assumed that stats compatibility between two types means that `toStatsRepresentation` returns the same result for a value with either Type. Please let me know if that definition is inaccurate.
A given input `value` can either be cast to a `long` for the "short" types or to a `Int128/LongTimestamp/LongTimestampWithTimeZone` for the "long" types in `toStatsRepresentation`. That's why I kept the conditions for both short or both long type and same scale.

File: core/trino-spi/src/main/java/io/trino/spi/statistics/StatsUtil.java [hunk 10]
```diff
@@ -85,4 +85,55 @@ public static OptionalDouble toStatsRepresentation(Type type, Object value)
 
         return OptionalDouble.empty();
     }
+
+    /**
+     * Stats representations of the given types are considered compatible if
+     * {@link #toStatsRepresentation} returns the same result for all valid values with either type.
```
Comment (findepi) [inline core/trino-spi/src/main/java/io/trino/spi/statistics/StatsUtil.java:91]: This isn't precise. Do you mean values that can be cast from one type to the other and back, without loss?

File: core/trino-spi/src/main/java/io/trino/spi/statistics/StatsUtil.java (hunk 7 above)
Comment (findepi) [reply core/trino-spi/src/main/java/io/trino/spi/statistics/StatsUtil.java:122]: > I've assumed that stats compatibility

Well, this shouldn't be "assumed". It should be "defined", as it's being added in this PR :) 

Anyway, whatever the definition is currently, it's not quite coherent yet. 
For example, you say that `integer` and `bigint` are compatible. `bigint` can represent all values of `integer` type and some more -- it's strictly wider and for the overlapping values, their stats representation is the same.

Following this example, `decimal(p, s)` and `decimal(p+1, s)` should be compatible -- the second one is strictly wider and for the overlapping values, their stats representation is the same.

TL;DR whatever the definition is going to be, coherence demands that there is no short/short or long/long condition here.

File: core/trino-spi/src/main/java/io/trino/spi/statistics/StatsUtil.java (hunk 10 above)
Comment (findepi) [reply core/trino-spi/src/main/java/io/trino/spi/statistics/StatsUtil.java:91]: Also, we should be clear whether we're defining an equivalence relation.

- Is this reflective? yes
- is this symmetric? currently it is (per how the method is defined & named; but maybe we want to model this around wider/narrower types and then it wouldn't be)
- is this transitive?

File: core/trino-main/src/main/java/io/trino/cost/FilterStatsCalculator.java [hunk 11]
```diff
@@ -105,6 +114,31 @@ public PlanNodeStatsEstimate filterStats(
...
+            return Optional.empty();
+        }
+        Cast castExpression = (Cast) expression;
+        if (!(castExpression.getExpression() instanceof SymbolReference)) {
+            return Optional.empty();
+        }
+        Map<NodeRef<Expression>, Type> expressionTypes = typeAnalyzer.getTypes(session, types, expression);
+        Type castSourceType = expressionTypes.get(NodeRef.of(castExpression.getExpression()));
+        Type castTargetType = expressionTypes.get(NodeRef.<Expression>of(castExpression));
+        // CAST must be an implicit coercion and the stats representation of target type
```
Comment (sopel39) [inline core/trino-main/src/main/java/io/trino/cost/FilterStatsCalculator.java:133]: > stats representation

You mean `SymbolStatsEstimate`?

> stats representation of target type should be compatible with the source type

I would say that target domain must be subset of source domain (e.g. `double -> bigint` cast). Is that the case?

Comment (findepi) [reply core/trino-main/src/main/java/io/trino/cost/FilterStatsCalculator.java:133]: > > stats representation
> 
> You mean `SymbolStatsEstimate`?

yes; precisely the `io.trino.spi.statistics.StatsUtil#toStatsRepresentation`

File: core/trino-main/src/main/java/io/trino/cost/FilterStatsCalculator.java [hunk 12]
```diff
@@ -105,6 +114,31 @@ public PlanNodeStatsEstimate filterStats(
                 .process(simplifiedExpression);
     }
 
+    @VisibleForTesting
+    Optional<Symbol> getSymbolReference(Expression expression, Session session, TypeProvider types)
```
Comment (findepi) [inline core/trino-main/src/main/java/io/trino/cost/FilterStatsCalculator.java:118]: `getSymbolReferenceForStats`

File: core/trino-main/src/main/java/io/trino/cost/FilterStatsCalculator.java [hunk 13]
```diff
@@ -105,6 +114,31 @@ public PlanNodeStatsEstimate filterStats(
...
+        if (!(expression instanceof Cast)) {
+            return Optional.empty();
+        }
+        Cast castExpression = (Cast) expression;
+        if (!(castExpression.getExpression() instanceof SymbolReference)) {
+            return Optional.empty();
+        }
+        Map<NodeRef<Expression>, Type> expressionTypes = typeAnalyzer.getTypes(session, types, expression);
+        Type castSourceType = expressionTypes.get(NodeRef.of(castExpression.getExpression()));
+        Type castTargetType = expressionTypes.get(NodeRef.<Expression>of(castExpression));
```
Comment (findepi) [inline core/trino-main/src/main/java/io/trino/cost/FilterStatsCalculator.java:132]: `<Expression>` is redundant

File: core/trino-main/src/main/java/io/trino/cost/FilterStatsCalculator.java (hunk 11 above)
Comment (findepi) [inline core/trino-main/src/main/java/io/trino/cost/FilterStatsCalculator.java:133]: > must be an implicit coercion

explain why.
we expect implicit coercion 

- to be generally injective... which is not true for coercion to double, so maybe worth also documenting why it is not a problem here. Compare with `io.trino.sql.planner.iterative.rule.UnwrapCastInComparison.Visitor#hasInjectiveImplicitCoercion` logic. I don't think we need to be as complex (this is stats estimates only), but we should document assumptions made
- to be order-preserving (see assumption documentation in `UnwrapCastInComparison` about "order-preserving")

File: core/trino-main/src/main/java/io/trino/cost/FilterStatsCalculator.java [hunk 14]
```diff
@@ -277,7 +311,8 @@ protected PlanNodeStatsEstimate visitIsNullPredicate(IsNullPredicate node, Void
         @Override
         protected PlanNodeStatsEstimate visitBetweenPredicate(BetweenPredicate node, Void context)
         {
-            if (!(node.getValue() instanceof SymbolReference)) {
+            SymbolStatsEstimate valueStats = getExpressionStats(node.getValue());
```
Comment (findepi) [inline core/trino-main/src/main/java/io/trino/cost/FilterStatsCalculator.java:314]: `FilterStatsCalculator` gains special logic to ignore coercions and thus produce better stats. Here, however, we're delegating to `scalarStatsCalculator`, which has different logic.

- in getExpressionStats, if `getSymbolReferenceForStats().isPresent()`, we should use this symbols stats
- in `scalarStatsCalculator` we should fix visitCast to honor `areStatsRepresentationsCompatible`

File: core/trino-main/src/main/java/io/trino/cost/FilterStatsCalculator.java (hunk 12 above)
Comment (findepi) [inline core/trino-main/src/main/java/io/trino/cost/FilterStatsCalculator.java:118]: move under `visitComparisonExpression` (or below all `visit*` methods)

**Review by findepi — CHANGES_REQUESTED** [review]

**Comment by findepi** [comment]
Added some more comments.
Seems my previous comments were not applied/discussed yet.
Let me know if any clarification is needed.

**Comment by bitsondatadev** [comment]
:wave: @raunaqmorarka  - this PR is inactive and doesn't seem to be under development. If you'd like to continue work on this at any point in the future, feel free to re-open.
